pkg/provider/provider.go

```diff
@@ -350,22 +403,83 @@ func (p *Provider) getKeyVaultObjectContent(ctx context.Context, kvClient *kv.Ba
 
 	switch kvObject.ObjectType {
 	case VaultObjectTypeSecret:
-		return getSecret(ctx, kvClient, vaultURL, kvObject)
+		return getSecret(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	case VaultObjectTypeKey:
-		return getKey(ctx, kvClient, vaultURL, kvObject)
+		return getKey(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	case VaultObjectTypeCertificate:
-		return getCertificate(ctx, kvClient, vaultURL, kvObject)
+		return getCertificate(ctx, kvClient, vaultURL, kvObject, objectVersion)
 	default:
 		err := errors.Errorf("Invalid vaultObjectTypes. Should be secret, key, or cert")
-		return "", "", wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+		return "", "", wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, objectVersion)
 	}
 }
 
+func (p *Provider) getKeyVaultSecretVersions(ctx context.Context, kvClient *kv.BaseClient, kvObject KeyVaultObject, vaultURL string) (versions []string, err error) {
+	start := time.Now()
+	defer func() {
+		var errMsg string
+		if err != nil {
+			errMsg = err.Error()
+		}
+		p.reporter.ReportKeyvaultRequest(ctx, time.Since(start).Seconds(), kvObject.ObjectType, kvObject.ObjectName, errMsg)
+	}()
+
+	// We have to pull all of the secret versions for a secret because we can't trust
+	// the values returned by KV to be in any reliable order
+	var maxResultsPerPage int32 = 25
+
+	kvVersionsList, err := kvClient.GetSecretVersions(ctx, vaultURL, kvObject.ObjectName, &maxResultsPerPage)
+	if err != nil {
+		return nil, wrapObjectTypeError(err, kvObject.ObjectType, kvObject.ObjectName, kvObject.ObjectVersion)
+	}
+
+	secretVersions := KeyVaultSecretList{}
+
+	for notDone := true; notDone; notDone = kvVersionsList.NotDone() {
+		for _, secret := range kvVersionsList.Values() {
+			objectVersion := getObjectVersion(*secret.ID)
+			created := date.UnixEpoch()
```

Comment:
Thank you for working on this,... sorry my **Go** level is `null`, how this `date` tie to `secret.Attributes.Created` :-)

Reply:
This initializes the created date as the Unix epoch in case secret.Attributes is null. A few lines below, its overriden to the created date of the secret if the attributes are present.

Reply:
Oh, I see... didn't look down :-) ... Thank you.

Honestly, I would prefer a failure if if such information didn't came.

---

Also, I remember at some point I mistakenly disabled a certificate version that was explicitly being referenced in the SecretProviderClass, that resulted in a failure, seemed like somewhere fetching the disabled certificated wasn't allow. So, maybe consider filtering out disable versions, if you are not already doing so.

If the `Go` SDK behaves similar to `az` or the PowerShell command, disable versions will come when listing all the versions.